File: src_py/decision_engine.py
"""
Project Limbo — SLA-Aware Decision & Action Layer
Evaluates prediction confidence scores and SLA timelines to execute automated actions:
Quiet Wait, Proactive Customer Notification, or Automated Fund Reversal.
"""

import time
import json
import threading
from datetime import datetime
from src_py import db

class DecisionEngine:
    def __init__(self):
        self.is_running = False
        self._thread = None

    def process_limbo_decisions(self):
        try:
            ambiguous_txns = db.fetch_all("SELECT * FROM transactions WHERE visible_status = 'pending' AND is_ambiguous = 1")
            now = datetime.utcnow()

            for txn in ambiguous_txns:
                score = txn["probability_score"]
                sla_deadline = datetime.fromisoformat(txn["sla_deadline"].replace('Z', ''))
                
                is_near_sla = (sla_deadline - now).total_seconds() < txn["expected_window_sec"]
                is_sla_breached = now >= sla_deadline

                if score >= 0.75:
                    if txn["action_taken"] != "WAIT_QUIETLY":
                        db.execute("UPDATE transactions SET action_taken = 'WAIT_QUIETLY' WHERE id = ?", (txn["id"],))
                        db.log_event(
                            txn["id"],
                            txn["merchant_id"],
                            "DECISION_WAIT_QUIETLY",
                            f"High confidence (P={score}). Engine waiting quietly for bank confirmation on {txn['id']}."
                        )
                elif score >= 0.35 and is_near_sla:
                    if txn["action_taken"] != "NOTIFY_CUSTOMER":
                        db.execute("UPDATE transactions SET action_taken = 'NOTIFY_CUSTOMER' WHERE id = ?", (txn["id"],))
                        meta = json.dumps({"probability": score, "amount": txn["amount"], "bank": txn["issuing_bank"]})
                        db.log_event(
                            txn["id"],
                            txn["merchant_id"],
                            "NOTIFICATION_SENT",
                            f"Proactive SMS/Email dispatched to Customer {txn['customer_id']}: Payment of ₹{txn['amount']} is debited and being tracked with {txn['issuing_bank']}. No re-payment needed.",
                            meta
                        )
                elif score < 0.35 and is_sla_breached:
                    if txn["action_taken"] != "AUTO_REVERSAL":
                        db.execute("UPDATE transactions SET visible_status = 'reversed', action_taken = 'AUTO_REVERSAL', is_ambiguous = 0 WHERE id = ?", (txn["id"],))
                        meta = json.dumps({"probability": score, "amount": txn["amount"], "bank": txn["issuing_bank"]})
                        db.log_event(
                            txn["id"],
                            txn["merchant_id"],
                            "REVERSAL_TRIGGERED",
                            f"Low confidence (P={score}) & SLA breached. Reversal triggered automatically for {txn['id']} (₹{txn['amount']}). Customer refunded.",
                            meta
                        )

        except Exception as e:
            print(f"Decision Engine error: {e}")

    def _loop(self, interval_sec: float = 2.5):
        while self.is_running:
            self.process_limbo_decisions()
            time.sleep(interval_sec)

    def start(self, interval_sec: float = 2.5):
        if self.is_running:
            return
        self.is_running = True
        self._thread = threading.Thread(target=self._loop, args=(interval_sec,), daemon=True)
        self._thread.start()
        print("Decision Engine started (Python background thread)")

    def stop(self):
        self.is_running = False
        print("Decision Engine stopped")

decision_engine = DecisionEngine()

File: src/engine/decision_engine.js
const db = require('../database/db');

class DecisionEngine {
  constructor() {
    this.isRunning = false;
    this.interval = null;
  }

  async processLimboDecisions() {
    try {
      const ambiguousTxns = await db.all(
        `SELECT * FROM transactions WHERE visible_status = 'pending' AND is_ambiguous = 1`
      );

      const now = new Date().getTime();

      for (const txn of ambiguousTxns) {
        const score = txn.probability_score;
        const slaDeadlineTime = new Date(txn.sla_deadline).getTime();
        const isNearSla = (slaDeadlineTime - now) < (txn.expected_window_sec * 1000);
        const isSlaBreached = now >= slaDeadlineTime;

        if (score >= 0.75) {
          // Action: Wait Quietly
          if (txn.action_taken !== 'WAIT_QUIETLY') {
            await db.run(
              `UPDATE transactions SET action_taken = 'WAIT_QUIETLY' WHERE id = ?`,
              [txn.id]
            );

            await db.run(
              `INSERT INTO event_logs (transaction_id, merchant_id, event_type, message)
               VALUES (?, ?, ?, ?)`,
              [
                txn.id,
                txn.merchant_id,
                'DECISION_WAIT_QUIETLY',
                `High confidence (P=${score}). Engine waiting quietly for bank confirmation on ${txn.id}.`
              ]
            );
          }
        } else if (score >= 0.35 && isNearSla) {
          // Action: Proactive Customer Notification
          if (txn.action_taken !== 'NOTIFY_CUSTOMER') {
            await db.run(
              `UPDATE transactions SET action_taken = 'NOTIFY_CUSTOMER' WHERE id = ?`,
              [txn.id]
            );

            await db.run(
              `INSERT INTO event_logs (transaction_id, merchant_id, event_type, message, metadata)
               VALUES (?, ?, ?, ?, ?)`,
              [
                txn.id,
                txn.merchant_id,
                'NOTIFICATION_SENT',
                `Proactive SMS/Email dispatched to Customer ${txn.customer_id}: Payment of ₹${txn.amount} is debited and being tracked with ${txn.issuing_bank}. No re-payment needed.`,
                JSON.stringify({ probability: score, amount: txn.amount, bank: txn.issuing_bank })
              ]
            );
          }
        } else if (score < 0.35 && isSlaBreached) {
          // Action: Trigger Automatic Reversal
          if (txn.action_taken !== 'AUTO_REVERSAL') {
            await db.run(
              `UPDATE transactions SET visible_status = 'reversed', action_taken = 'AUTO_REVERSAL', is_ambiguous = 0 WHERE id = ?`,
              [txn.id]
            );

            await db.run(
              `INSERT INTO event_logs (transaction_id, merchant_id, event_type, message, metadata)
               VALUES (?, ?, ?, ?, ?)`,
              [
                txn.id,
                txn.merchant_id,
                'REVERSAL_TRIGGERED',
                `Low confidence (P=${score}) & SLA breached. Reversal triggered automatically for ${txn.id} (₹${txn.amount}). Customer refunded.`,
                JSON.stringify({ probability: score, amount: txn.amount, bank: txn.issuing_bank })
              ]
            );
          }
        }
      }
    } catch (err) {
      console.error('Decision Engine processing error:', err.message);
    }
  }

  start(intervalMs = 2500) {
    if (this.isRunning) return;
    this.isRunning = true;
    console.log(`Decision Engine started (Interval: ${intervalMs}ms)`);

    this.interval = setInterval(() => {
      this.processLimboDecisions();
    }, intervalMs);
  }

  stop() {
    if (!this.isRunning) return;
    clearInterval(this.interval);
    this.isRunning = false;
    console.log('Decision Engine stopped');
  }
}

module.exports = new DecisionEngine();
